contents/security/crossSiteRequestForgery/crossSiteRequestForgery.md

```diff
@@ -0,0 +1,96 @@
+
+Cross-Site Request Forgery (CSRF) is a dangerous type of attack that has affected major sites like [Gmail](http://archive.oreilly.com/pub/post/gmail_exploit_contact_list_hij.html) or [Netflix](http://blog.jeremiahgrossman.com/2006/10/more-on-netflixs-csrf-advisory.html) in the past. This article attempts to give an easy-to-digest introduction to the attack and how to protect your website from it.
+
+## Flow of a CSRF attack
+
+A CSRF attack tricks the victim to perform actions that they do not intend to do on a web application in which they're currently authenticated. It is best explained through an example. Suppose that Alice is a customer of the banking website `www.example-bank.com`. Alice wants to transfer money to her friend Bob (because she does not like the fact that he keeps paying for her meals when they go out together). 
+
+Alice first logs in to the website at `www.example-bank.com/login`, providing her username and password. She then clicks on a hyperlink on the site to go to the URL `www.example-bank.com/transfer` to perform her transaction. Note that Alice does not have to enter her username and password again, but the bank still knows that the request is made by her. This is done through a mechanism of the HTTP protocol called a [HTTP Cookie](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies). A *cookie* is a small piece of information that is sent along with all HTTP requests to a particular website. In our case, after receiving Alice's credentials for logging in, the bank instructs Alice's browser to store a *cookie* `auth=1abcd2ek3292fsa390sdf` and **send this cookie together with every subsequent requests** to `www.example-bank.com`. Thus, once the web server sees this cookie in the request, it knows that this is an authenticated request from Alice. Many websites use this cookie mechanism so that users only have to log in once.
+
+At `www.example-bank.com/transfer`, Alice fills out the following HTML form to transfer money to Bob:
+
+```
+<form action="/transfer" method="POST">
+<input type="text" name="receiver-name" value="Bob"/>
+<input type="number" name="receiver-account-no" value="123456"/>
+<input type="number" name="amount" value="100"/>
+<input type="submit"/>
+</form>
+```
+
+As Alice clicks the submit button, the browser will send the following request to `www.example-bank.com`:
+
+```
+POST http://www.example-bank.com/transfer HTTP/1.1
+Cookie: auth=1abcd2ek3292fsa390sdf
+
+receiver-name=Bob&receiver-account-no=123456&amount=100
+```
+
+The bank's server verifies that the cookie `auth=1abcd2ek3292fsa390sdf` is valid and is associated with Alice. It then proceeds to transfer $100 from Alice's account to account number 123456, which belongs to Bob.
+
+Alice then receives a chat message from her not-so-trustworthy friend, Eve:
+
+```
+Eve: OMG Alice! This site is selling the shoes you have always wanted for half the price: www.i-am-not-evil.com/shoes
```

Comment:
Consider adding a line break here? The code snippet requires using a horizontal scroll to finish reading even when fullscreen